python/k4FWCore/ApplicationMgr.py

```diff
@@ -0,0 +1,34 @@
+#
+# Copyright (c) 2014-2024 Key4hep-Project.
+#
+# This file is part of Key4hep.
+# See https://key4hep.github.io/key4hep-doc/ for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from Configurables import ApplicationMgr as AppMgr
+from Configurables import Reader, Writer, IOSvc
+from Configurables import Gaudi__Sequencer
+
+class ApplicationMgr:
+
+    def __init__(self, **kwargs):
+        self._mgr = AppMgr(**kwargs)
+
+        for conf in frozenset(self._mgr.allConfigurables.values()):
+            if isinstance(conf, IOSvc):
+                props = conf.getPropertiesWithDescription()
+                if 'input' in props:
+                    self._mgr.TopAlg =  [Reader("Reader")] + self._mgr.TopAlg
+                if 'output' in props:
+                    self._mgr.TopAlg = [Gaudi__Sequencer("Sequencer", Members=[*self._mgr.TopAlg, Writer("Writer")], Sequential=True)]
```

Comment:
Does it mean it's not allowed to put `Reader` and `Writer` explicitly in the  `topAlgs`, because that would create two instances?

To avoid name collision maybe a more detail name for Sequencer could be used, e.g. 'IOSequencer' or  'IOTopLevelSequencer' , something with 'default' or '_'

I guess the sequential sequencer is used to enforce that the `Reader`, client algorithms and `Writer`  happens in that order, which otherwise is not guaranteed since read/write have no data dependency.
- in case when 'input' is present but not 'output', shouldn't the `topAlgs` be wrapped in a sequencer too?
- I believe the top level algorithms are normally operated in a concurrent mode. After wrapping everything in a sequential sequencer the client algorithms will be executed in sequential manner. In order to have a sequence  `Reader`, client algorithms, `Writer` but allow concurrent execution within client algorithms, the client algorithms could be wrapped in yet another sequencer:
 ```python
self._mgr.TopAlg = [
    Gaudi__Sequencer(
        "Sequencer",
        Sequential=True,
        Members=[
            Reader("Reader"),
            Gaudi__Sequencer(
                "InnerSequencer", Sequential=False, Members=self._mgr.TopAlg
            ),
            Writer("Writer"),
        ],
    )
]
```

Please correct me if I'm wrong

Reply:
A few things have changed since this comment:
- `Reader` and `Writer` can be in `TopAlg`, then they won't be added, but if they are not and an input file or output file is added, then they will be added
- `Reader` (if any) and the algorithms are wrapped together since the `Reader` has information about which collections it is going to read and can be scheduled fine, the writer is wrapped with `Sequential=False` similarly to the snippet above.

Reply:
> A few things have changed since this comment:
> 
>     * `Reader` and `Writer` can be in `TopAlg`, then they won't be added, but if they are not and an input file or output file is added, then they will be added
> 
>     * `Reader` (if any) and the algorithms are wrapped together since the `Reader` has information about which collections it is going to read and can be scheduled fine, the writer is wrapped with `Sequential=False` similarly to the snippet above.
>       I tested this setup and the algorithms can run in parallel while the Writer runs last

Can you put this into the docstring?